blaze/compute/numba.py

```diff
@@ -0,0 +1,203 @@
+        restype = numba.types.NPDatetime('us')
+    elif isinstance(measure, datashape.TimeDelta):  # isinstance for diff freqs
+        restype = numba.types.NPTimedelta(measure.unit)
+    elif isinstance(measure, datashape.String):
+        encoding = measure.encoding
+        fixlen = measure.fixlen
+        if fixlen is None:
+            if encoding == 'A':
+                return numba.types.string
+            raise TypeError("Numba cannot handle variable length strings")
+        typ = (numba.types.CharSeq
+               if encoding == 'A' else numba.types.UnicodeCharSeq)
+        return typ(fixlen or 0)
+    elif measure == datashape.object_:
+        raise TypeError("Numba cannot handle object datashape")
+    else:
+        restype = getattr(numba, str(measure))
+    return restype
+
+
+def compute_signature(expr):
+    """Get the ``numba`` *function signature* corresponding to ``DataShape``
+
+    Examples
+    --------
+    >>> from blaze import symbol
+    >>> s = symbol('s', 'int64')
+    >>> t = symbol('t', 'float32')
+    >>> d = symbol('d', 'datetime')
+
+    >>> expr = s + t
+    >>> compute_signature(expr)
+    float64(int64, float32)
+
+    >>> expr = d.truncate(days=1)
+    >>> compute_signature(expr)
+    datetime64(D)(datetime64(us))
+
+    >>> expr = d.day + 1
+    >>> compute_signature(expr)  # only looks at leaf nodes
+    int64(datetime64(us))
+
+    Notes
+    -----
+    * This could potentially be adapted/refactored to deal with
+      ``datashape.Function`` types.
+    * Cannot handle ``datashape.Record`` types.
+    """
+    assert datashape.isscalar(expr.schema)
+    restype = get_numba_type(expr.schema)
+    argtypes = [get_numba_type(e.schema) for e in expr._leaves()]
+    return restype(*argtypes)
+
+
+def _get_numba_ufunc(expr):
+    """Construct a numba ufunc from a blaze expression
+
+    Parameters
+    ----------
+    expr : blaze.expr.Expr
+
+    Returns
+    -------
+    f : function
+        A numba vectorized function
+
+    Examples
+    --------
+    >>> from blaze import symbol
+    >>> import numpy as np
+
+    >>> s = symbol('s', 'float64')
+    >>> t = symbol('t', 'float64')
+
+    >>> x = np.array([1.0, 2.0, 3.0])
+    >>> y = np.array([2.0, 3.0, 4.0])
+
+    >>> f = get_numba_ufunc(s + t)
+
+    >>> f(x, y)
+    array([ 3.,  5.,  7.])
+
+    See Also
+    --------
+    get_numba_type
+    compute_signature
+    """
+    leaves = expr._leaves()
+    s, scope = funcstr(leaves, expr)
+    scope = dict((k, numba.jit(nopython=True)(v) if callable(v) else v)
+                 for k, v in scope.items())
+    func = eval(s, scope)
+    sig = compute_signature(expr)
+    return numba.vectorize([sig], nopython=True)(func)
+
+
+# do this here so we can run our doctest
+get_numba_ufunc = memoize(_get_numba_ufunc)
+
+
+@dispatch(Broadcast, np.ndarray)
+def compute_up(t, x, **kwargs):
+    assert len(t._scalars) == 1
+    scalar = t._scalars[0]
+    fields = scalar.fields
```

Comment:
Hrm, we should also test this with non-record arrays.  Datashapes like `10 * 10 * int` will also be a common use case.